src/sciline/container.py

```diff
@@ -14,81 +25,140 @@ class UnsatisfiedRequirement(Exception):
-    parameters injected.
-    """
-    # When building a workflow, there are two common problems:
-    #
-    # 1. Intermediate results are used more than once.
-    # 2. Intermediate results are large, so we generally do not want to keep them
-    #    in memory longer than necessary.
-    #
-    # To address these problems, we can internally build a graph of tasks, instead of
-    # directly creating dependencies between functions. Currently we use Dask for this.
-    # The Container instance will automatically compute the task, unless it is marked
-    # as lazy. We therefore use singleton-scope (to ensure Dask will recognize the
-    # task as the same object) and also wrap the function in dask.delayed.
-    scope = injector.singleton
-    func = _delayed(func)
-    tps = typing.get_type_hints(func)
-
-    def bind(binder: injector.Binder):
-        binder.bind(tps['return'], injector.inject(func), scope=scope)
-
-    return bind
-
-
-def make_container(funcs: List[Callable], /, *, lazy: bool = False) -> Container:
-    """
-    Create a :py:class:`Container` from a list of functions.
-
-    Parameters
-    ----------
-    funcs:
-        List of functions to be injected. Must be annotated with type hints.
-    lazy:
-        If True, the functions are wrapped in :py:func:`dask.delayed` before
-        being injected. This allows to build a dask graph from the container.
-    """
-    # Note that we disable auto_bind, to ensure we do not accidentally bind to
-    # some default values. Everything must be explicit.
-    return Container(
-        injector.Injector([_injectable(f) for f in funcs], auto_bind=False),
-        lazy=lazy,
-    )
+Provider = Callable[..., Any]
+
+
+class Container:
+    def __init__(self, funcs: List[Provider], /):
+        """
+        Create a :py:class:`Container` from a list of functions.
+
+        Parameters
+        ----------
+        funcs:
+            List of functions to be injected. Must be annotated with type hints.
+        """
+        self._providers: Dict[type, Provider] = {}
+        self._subproviders: Dict[type, Dict[Tuple[type | TypeVar, ...], Provider]] = {}
+        self._cache: Dict[type, Delayed] = {}
+        for func in funcs:
+            self.insert(func)
+
+    def insert(self, provider: Provider) -> None:
+        if (key := get_type_hints(provider).get('return')) is None:
+            raise ValueError(f'Provider {provider} lacks type-hint for return value')
+        # isinstance does not work here and types.NoneType available only in 3.10+
+        if key == type(None):  # noqa: E721
+            raise ValueError(f'Provider {provider} returning `None` is not allowed')
+        if (origin := get_origin(key)) is not None:
+            subproviders = self._subproviders.setdefault(origin, {})
+            args = get_args(key)
+            if args in subproviders:
+                raise ValueError(f'Provider for {key} already exists')
+            subproviders[args] = provider
```

Comment:
Should we forbid this kind of nested generic arguments...?

```
VT = TypeVar("VT")
Bike = NewType("Bike", object)
Car = NewType("Car", object)

class Time(sl.Scope[VT], float): ...

def provide_bike_time() -> Time[Bike]:
  return Time(1)
def provide_car_time() -> Time[Car]:
  return Time(0.1)

def provide_list_of_time(bike_time: Time[VT]) -> List[Time[VT]]
  return [bike_time]
```

It should be 
```
TimeT = TypeVar("TimeT")
def provide_list_of_time(bike_time: TimeT) -> List[TimeT]
  return [bike_time]
```
And then we can compute nested generic type like; `List[Time[Car]]` or `List[Time[Bike]]`.